### PR TITLE
First steps in replacing `KClass` with custom information type

### LINKED ISSUE #112: Consider replacing `KClass` with custom information class as DSL result
The idea is that `WithX.toList` should return not a `KClass`, which assumes further runtime reflection, but an information class storing all the information about the class available to Reflekt plus the old `KClass` reference.

This is particularly useful to implement #71, because actually `KClass` has the `sealedSubclasses` property that does that thing; however, this data can be collected at compile time, and it's helpful in cases where run-time reflection isn't available or too slow.

Rough API:

```kotlin
class ReflektClass<T> {
  val kotlin: KClass<T>
  val qualifiledName: String
  val simpleName: String
  val sealedSubclasses: List<ReflektClass<*>>
  val isOpen: Boolean
  ...
}
```

Possible usage:
```kotlin
Reflekt.classes().withSuperType<B1>().toList().map { /* it: ReflektClass<B1> */ it.qualifiedName }.forEach { println(it) }
```

Also, to not bloat the `ReflektImpl` file, there must be a registry of such classes to have them available in many DSL calls simultaneously (basically, `val classData: HashMap<String, ReflektClass<*>>`).

### PR DESCRIPTION
**Pull request by CommanderTvis**
Very first steps in replacing `KClass` as query result with a custom information class `ReflectClass`.

Sorry for lots of unrelated refactorings. I can extract them to another PR if needed.

Resolves #112

### CODE REVIEW

File: reflekt-dsl/src/main/kotlin/org/jetbrains/reflekt/ReflektClass.kt [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ * Represents a class and designed for storing information about it, which is collected by Reflekt at compile-time.
+ *
+ * @param T the type of the class.
+ * @property kotlin [KClass] instance for this class.
+ * @property annotations Annotations which are present on this class.
+ * @property isAbstract `true` if this class is `abstract`.
+ * @property isCompanion `true` if this class is a companion object.
+ * @property isData `true` if this class is a data class.
+ * @property isFinal `true` if this class is `final`.
+ * @property isFun `true` if this class is a Kotlin functional interface.
```
Comment (nbirillo) [inline reflekt-dsl/src/main/kotlin/org/jetbrains/reflekt/ReflektClass.kt:15]: Maybe `isSam` is better? But I am not sure

File: reflekt-dsl/src/main/kotlin/org/jetbrains/reflekt/ReflektClass.kt [hunk 2]
```diff
@@ -0,0 +1,44 @@
+package org.jetbrains.reflekt
+
+import kotlin.reflect.*
+
+/**
+ * Represents a class and designed for storing information about it, which is collected by Reflekt at compile-time.
+ *
+ * @param T the type of the class.
+ * @property kotlin [KClass] instance for this class.
```
Comment (nbirillo) [inline reflekt-dsl/src/main/kotlin/org/jetbrains/reflekt/ReflektClass.kt:9]: `kClass`? `kotlin` looks a little bit strange for me..

File: reflekt-dsl/src/main/kotlin/org/jetbrains/reflekt/ReflektClass.kt [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+ * @property annotations Annotations which are present on this class.
```
Comment (nbirillo) [inline reflekt-dsl/src/main/kotlin/org/jetbrains/reflekt/ReflektClass.kt:10]: Are we going to store `qualifiedNames`, right?

File: reflekt-dsl/src/main/kotlin/org/jetbrains/reflekt/ReflektClass.kt [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+ * @property isSealed `true` if this class is `sealed`.
+ * @property isValue `true` if this class is a value class.
+ * @property qualifiedName The fully qualified dot-separated name of the class, or `null` if the class is local or a class of an anonymous object.
+ * @property superclasses The set of immediate superclasses of this class.
+ * @property sealedSubclasses The set of the immediate subclasses if this class is a sealed class, or an empty set otherwise.
+ * @property simpleName The simple name of the class as it was declared in the source code,
+ * or `null` if the class has no name (if, for example, it is a class of an anonymous object).
+ * @property visibility Visibility of this class, or `null` if its visibility cannot be represented in Kotlin.
+ */
+data class ReflektClass<T : Any>(
```
Comment (nbirillo) [inline reflekt-dsl/src/main/kotlin/org/jetbrains/reflekt/ReflektClass.kt:27]: What is about functions? I mean e.g. receivers

**Comment by nbirillo** [comment]
Also I think to extract refactorings into another PR is a great idea

File: reflekt-dsl/src/main/kotlin/org/jetbrains/reflekt/ReflektClass.kt (hunk 2 above)
Comment (CommanderTvis) [reply reflekt-dsl/src/main/kotlin/org/jetbrains/reflekt/ReflektClass.kt:9]: OK

File: reflekt-dsl/src/main/kotlin/org/jetbrains/reflekt/ReflektClass.kt (hunk 4 above)
Comment (CommanderTvis) [reply reflekt-dsl/src/main/kotlin/org/jetbrains/reflekt/ReflektClass.kt:27]: In this PR I am going only to have info for classes.

**Pull request closed without merging**